PCF-ADR-0004-codeflare-test-strategy.md

```diff
@@ -0,0 +1,392 @@
+* Focusing on functional testing, avoid long running tests with lower value (tested by nightly checks)
+
+#### Testing levels and Type of testing:
+
+| Test type/development cycle  | Feature developer | PR automated checks  | PR review by reviewer | Nightly automated checks  | Release automated checks |
+| ---------------------------- | ----------------- | -------------------- | --------------------- | ------------------------- | ------------------------ |
+| Manual testing               | Yes               | No                   | Yes                   | No                        | No                       |
+| Unit testing                 | Yes               | Yes                  | No                    | Yes                       | Yes                      |
+| End to end testing           | Yes               | Yes                  | No                    | Yes                       | Yes                      |
+| Integration testing          | Optional          | Yes                  | No                    | Yes                       | Yes                      |
+| Upgrade testing              | Optional          | Yes                  | No                    | Yes                       | Yes                      |
+
+Glossary:
+
+Manual testing	- Manual verification of the functionality.
+
+Unit testing	- Unit tests covering the base functionality.
+
+End to end testing 	- Testing including all CodeFlare components, running in a cloud.
+
+Integration testing 	- Testing of interactions between ODH components and CodeFlare components in a cloud.
+
+Upgrade testing	- Testing of component upgrades (deploy new operator version) in a cloud.
+
+### Test Environment
+Test environment depends on resource requirements of execued tests. For unit tests we will leverage GitHub actions runner. For end to end/integration/upgrade testing we will currently use KinD cluster running on GitHub actions runner. In case the resources provided by GitHub actions runner are not sufficient we can consider alternatives like [testing farm](https://docs.testing-farm.io/general/0.1/index.html).
+
+### Defect Management 
+
+#### Bug logging
+Bugs found by testing are reported as GitHub issues in respective repository. Every bug description has to contain steps to reproduce.
+
+#### Bug fixing
+Bugs are going to be fixed in a similar approach as new features - planned as part of sprint planning, fixed through PRs. If possible the fix contains test coverage, preventing the issue from happening again.
+
+
+### Test Suite
+
+#### E2E Test Cases
+
+##### Setup
+
+The following steps must be executed once, with a cluster admin role, as prerequisites to running the e2e test cases:
+1. Provision a test Kubernetes cluster (resp. an OpenShift cluster)
+2. Build the CodeFlare operator container image:
+  * Clone the CodeFlare operator source code repository
+  * Checkout the branch to be tested, e.g., the PR feature branch
```

Comment:
These can be combined with something like:
```
git clone --branch mytag0.1 --depth 1 https://example.com/my/repo.git
```
(note the `--branch` option is a bit misleading as it accepts any git ref)

Reply:
Added a note with the git command example